// live-mutex/dist/main.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.r2gSmokeTest = exports.lmUtils = void 0;
const lmUtils = require("./utils");
exports.lmUtils = lmUtils;
var rw_client_1 = require("./rw-client");
Object.defineProperty(exports, "RWLockClient", { enumerable: true, get: function () { return rw_client_1.RWLockClient; } });
Object.defineProperty(exports, "RWLockReadPrefClient", { enumerable: true, get: function () { return rw_client_1.RWLockReadPrefClient; } });
var rw_write_preferred_client_1 = require("./rw-write-preferred-client");
Object.defineProperty(exports, "RWLockWritePrefClient", { enumerable: true, get: function () { return rw_write_preferred_client_1.RWLockWritePrefClient; } });
var client_1 = require("./client");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return client_1.Client; } });
Object.defineProperty(exports, "LMXClient", { enumerable: true, get: function () { return client_1.LMXClient; } });
Object.defineProperty(exports, "LvMtxClient", { enumerable: true, get: function () { return client_1.LvMtxClient; } });
var broker_1 = require("./broker");
Object.defineProperty(exports, "Broker", { enumerable: true, get: function () { return broker_1.Broker; } });
Object.defineProperty(exports, "LMXBroker", { enumerable: true, get: function () { return broker_1.LMXBroker; } });
Object.defineProperty(exports, "LvMtxBroker", { enumerable: true, get: function () { return broker_1.LvMtxBroker; } });
var shared_internal_1 = require("./shared-internal");
Object.defineProperty(exports, "LMXLockRequestError", { enumerable: true, get: function () { return shared_internal_1.LMXLockRequestError; } });
Object.defineProperty(exports, "LMXUnlockRequestError", { enumerable: true, get: function () { return shared_internal_1.LMXUnlockRequestError; } });
exports.r2gSmokeTest = function () {
    return true;
};
